File: src/nl/rubensten/texifyidea/insight/LatexParameterInfoHandler.java
package nl.rubensten.texifyidea.insight;

import com.intellij.codeInsight.lookup.LookupElement;
import com.intellij.lang.parameterInfo.*;
import com.intellij.psi.PsiElement;
import com.intellij.psi.PsiFile;
import com.intellij.psi.util.PsiTreeUtil;
import com.intellij.util.ArrayUtil;
import nl.rubensten.texifyidea.lang.Argument;
import nl.rubensten.texifyidea.lang.LatexNoMathCommand;
import nl.rubensten.texifyidea.psi.LatexCommands;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.Optional;

/**
 * @author Sten Wessel
 */
public class LatexParameterInfoHandler implements ParameterInfoHandler<LatexCommands, LatexNoMathCommand> {

    @Nullable
    private static LatexCommands findLatexCommand(PsiFile file, int offset) {
        PsiElement element = file.findElementAt(offset);
        return PsiTreeUtil.getParentOfType(element, LatexCommands.class);
    }

    @Override
    public boolean couldShowInLookup() {
        return true;
    }

    @Nullable
    @Override
    public Object[] getParametersForLookup(LookupElement item, ParameterInfoContext context) {
        return ArrayUtil.EMPTY_OBJECT_ARRAY;
    }

    @Nullable
    @Override
    public Object[] getParametersForDocumentation(LatexNoMathCommand p, ParameterInfoContext context) {
        return ArrayUtil.EMPTY_OBJECT_ARRAY;
    }

    @Nullable
    @Override
    public LatexCommands findElementForParameterInfo(@NotNull CreateParameterInfoContext context) {
        return findLatexCommand(context.getFile(), context.getOffset());
    }

    @Override
    public void showParameterInfo(@NotNull LatexCommands element, @NotNull CreateParameterInfoContext context) {
        Optional<LatexNoMathCommand> commandHuh =
                LatexNoMathCommand.get(element.getCommandToken().getText().substring(1));
        if (!commandHuh.isPresent()) {
            return;
        }

        context.setItemsToShow(new Object[] { commandHuh.get() });
        context.showHint(element, element.getTextOffset(), this);
    }

    @Nullable
    @Override
    public LatexCommands findElementForUpdatingParameterInfo(@NotNull UpdateParameterInfoContext
                                                                     context) {
        return findLatexCommand(context.getFile(), context.getOffset());
    }

    @Override
    public void updateParameterInfo(@NotNull LatexCommands element, @NotNull
            UpdateParameterInfoContext context) {
        context.setCurrentParameter(0);
    }

    @Nullable
    @Override
    public String getParameterCloseChars() {
        return "]}";
    }

    @Override
    public boolean tracksParameterIndex() {
        return true;
    }

    @Override
    public void updateUI(LatexNoMathCommand cmd, @NotNull ParameterInfoUIContext context) {
        if (cmd == null) {
            context.setUIComponentEnabled(false);
            return;
        }

        int index = context.getCurrentParameterIndex();
        Argument[] arguments = cmd.getArguments();

        if (index >= arguments.length) {
            context.setUIComponentEnabled(false);
            return;
        }

        context.setupUIComponentPresentation(cmd.getCommandDisplay() + cmd.getArgumentsDisplay(), 0, 0, false, false, true, context.getDefaultParameterColor());
    }
}

File: src/nl/hannahsten/texifyidea/insight/LatexParameterInfoHandler.kt
package nl.hannahsten.texifyidea.insight

import com.intellij.codeInsight.lookup.LookupElement
import com.intellij.lang.parameterInfo.*
import com.intellij.psi.PsiFile
import com.intellij.psi.util.PsiTreeUtil
import com.intellij.util.ArrayUtil
import nl.hannahsten.texifyidea.lang.LatexRegularCommand
import nl.hannahsten.texifyidea.psi.LatexCommands

/**
 * @author Sten Wessel
 */
class LatexParameterInfoHandler : ParameterInfoHandler<LatexCommands, LatexRegularCommand> {

    private fun findLatexCommand(file: PsiFile, offset: Int): LatexCommands? {
        val element = file.findElementAt(offset)
        return PsiTreeUtil.getParentOfType(element, LatexCommands::class.java)
    }

    override fun couldShowInLookup() = true

    override fun getParametersForLookup(item: LookupElement, context: ParameterInfoContext): Array<Any>? {
        return ArrayUtil.EMPTY_OBJECT_ARRAY
    }

    override fun findElementForParameterInfo(context: CreateParameterInfoContext): LatexCommands? {
        return findLatexCommand(context.file, context.offset)
    }

    override fun showParameterInfo(element: LatexCommands, context: CreateParameterInfoContext) {
        val commandHuh = LatexRegularCommand[element.commandToken.text.substring(1)] ?: return

        context.itemsToShow = arrayOf<Any>(commandHuh)
        context.showHint(element, element.textOffset, this)
    }

    override fun findElementForUpdatingParameterInfo(context: UpdateParameterInfoContext): LatexCommands? {
        return findLatexCommand(context.file, context.offset)
    }

    override fun updateParameterInfo(element: LatexCommands, context: UpdateParameterInfoContext) {
        context.setCurrentParameter(0)
    }

    override fun updateUI(cmd: LatexRegularCommand?, context: ParameterInfoUIContext) {
        if (cmd == null) {
            context.isUIComponentEnabled = false
            return
        }

        val index = context.currentParameterIndex
        val arguments = cmd.arguments

        if (index >= arguments.size) {
            context.isUIComponentEnabled = false
            return
        }

        context.setupUIComponentPresentation(cmd.commandDisplay + cmd.getArgumentsDisplay(), 0, 0, false, false, true, context.defaultParameterColor)
    }
}
